[TwitterBridge] Migration to API V1.1

arnd-s opened the pull request:
[TwitterBridge] Migration to API V1.1

Migrate all API calls to V1.1 This is an intermediate PR, so the bridge will be working again.

Final goal should be a move to V2.0

arnd-s committed:
[TwitterBridge] Migration to API V1.1 - Get Tweets by username / first step

arnd-s commented:
Currently only search by username without filters is migrated.

Also the PR currently contains garbage and debug elements.

arnd-s committed:
Fixed Retweets

arnd-s committed:
Commented out some old code without function and fixed json object name for text attribute of tweet

arnd-s committed:
Fix phpcs errors

arnd-s committed:
Switch to extended Tweet format to get full text again

arnd-s committed:
Get the full text from retweeted_status element

csisoap commented:
I hope V2 will have a way to filter out Pinned Tweet, Promo Tweet and any Reply to another user. I'm only want self-Reply from that user and old PR has done a great job from doing this.

arnd-s commented:
> I hope V2 will have a way to filter out Pinned Tweet, Promo Tweet and any Reply to another user. I'm only want self-Reply from that user and old PR has done a great job from doing this.

I will later add this filter again. It's currently commented out until i matched everything back to tweet structure from V1.1
V2 will be a separate PR later when everything is ported to at least 1.1. Primary goal is, to have it running ;)

Looks like for V2 authentication has to be changed again, based on some quick tests i made

toelke commented:
FWIW, I am running this branch for a few hours and it is way more "stable" than `main`. I now got error 401 for one feed. I will keep monitoring.

arnd-s committed:
In usersearch mode, load userdata to hide pinned tweets.

Added API call to fetch user object.

arnd-s committed:
start to restructure collectData function for cleaning up code, also added generic makeApiCall funtion

arnd-s committed:
Migrated "by keyword or hashtag' context

arnd-s committed:
Reactivated filter for replies in "by username" context.

Also commented out some not anymore used functions

arnd-s committed:
Fix phpcs error

arnd-s commented:
@anyone using this branch: It would be nice go get some feedback about whats broken or changed on the output side. For some cases I'm missing data to compare it with past behavior and I'm also missing old API documentation. So I'm currently trying my best to match everything from reading the old code against official API documentation.
Also general change requests or additions are welcomed and i will try my best to integrate them.

linoth commented:
Disclaimer: Not trying to advocate for one behavior over another, just trying to list differences that I've noticed.

- Currently, retweets link to what I guess is the action of retweeting?, rather than the original tweet.  For example, the feed for @\Twitter now links to: https://twitter.com/TwitterBlue/status/1484314336772321282

which Twitter then resolves to: https://twitter.com/Twitter/status/1484314336772321282 ("Twitter Retweeted foo")

Previous behavior would have been to just link straight to: https://twitter.com/TwitterBlue/status/1484226494708662273

The current behavior also won't show replies when you open the URL unless you click on the tweet's time to go to the tweet itself.  This behavior also impacts any feed reader that used URL to detect duplicate entries, for better or worse.  Username behavior has also changed, and would have been "TwitterBlue (@\TwitterBlue) RT: @\Twitter"

- Possibly an artifact of the above, the timestamp seems like it may be when the user took the action of retweeting, rather than the timestamp of the original tweet.  Using the example, 2022-01-20T23:58:09+00:00 is used instead of 2022-01-20T18:09:??+00:00.

- Replies are now present, but you've also already pointed out that is temporary.

- I feel like image resizing just defaults to "orig" which I have no problem with, but that previously required noimgscaling to be set, and I'm sure others' use case may make that not ideal.

I'm sleep deprived and that's everything that comes to mind at the moment.  Thank you for picking up this torch and running hard with it from the get-go.

Edit: corrected username behavior and attempted to remove account pinging.

arnd-s committed:
Store apikey and guesttoken inside class instance to save some cycles

arnd-s committed:
Updated tweet filtering to hide replies independend from calling context

arnd-s committed:
On retweets take timestamp from original tweet instead of when it was retweeted

arnd-s committed:
fix phpcs errors

arnd-s commented:
@linoth Thanks for the input

- replies should now be filtered if norep parameter was given. I think, based on old code, filtering out replies was always the case in the past but i thought it would be better to use the already available parameter for this.
- timestamp should now be from original tweet instead of the retweet.

mnalis commented:
hm, doesn't seem to work for me at all. I get latest one just now from https://raw.githubusercontent.com/arnd-s/rss-bridge/migrate_to_api_1.1/bridges/TwitterBridge.php and replace original one from `2022-01-20+dfsg1-1` with it, and search only for username without adding any checkboxes. I get this error:

```
Bridge returned error 403! (19017)
2022-01-25T13:09:21+00:00

Twitter @elonmusk was unable to receive or process the remote website's content!
Error message: `Unexpected response from upstream.
cUrl error: (0)
PHP error: `
Query string: `action=display&bridge=Twitter&context=By+username&u=elonmusk&format=Html`
Version: `dev.2022-01-20`

    Press Return to check your input parameters
    Press F5 to retry
    Check if this issue was already reported on GitHub (give it a thumbs-up)
    Open a GitHub Issue if this error persists
```

**UPDATE**: Minor downtimes of few minutes and restarts didn't help. But after disabling rss-bridge twitter bridge for half a day, and enabling it again, now it seems to work ok!

em92 commented:
Yesterday evening deployed on https://feed.eugenemolotov.ru/pr2433/
In logs I got 2 times this:
```
[26-Jan-2022 01:41:21 UTC] Exception: Unexpected response from upstream.
cUrl error:  (0)
PHP error:  in /var/www/rss-bridge/pr2433/lib/error.php:24
Stack trace:
#0 /var/www/rss-bridge/pr2433/lib/contents.php(203): returnError()
#1 /var/www/rss-bridge/pr2433/bridges/TwitterBridge.php(735): getContents()
#2 /var/www/rss-bridge/pr2433/bridges/TwitterBridge.php(252): TwitterBridge->makeApiCall()
#3 /var/www/rss-bridge/pr2433/actions/DisplayAction.php(137): TwitterBridge->collectData()
#4 /var/www/rss-bridge/pr2433/index.php(38): DisplayAction->execute()
#5 {main}
```
My feed reader uses one feed on that deployment. 40 minutes between queries.

arnd-s commented:
@mnalis This looks like expired/invalid authentication information. I will rewrite some code today and try to catch this kind of error or at least return a more informative error message to solve this.

@em92 Do you got also 403 errors ?

@linoth the noimgscaling option has no use. This was also the case for old code before i started the rewrites. The parts for embedding images wasn't touched by me so far.
Links for retweets will be fixed today

arnd-s committed:
Fix link to tweet in case of a retweet

arnd-s committed:
Filterout retweets only for search by hashtag or key

arnd-s committed:
fix bug with continue not working in case of filtering out a tweet

arnd-s committed:
Rewrote Authcode

getapikey will normaly now only called once and store all required data inside class object.

When an API call fails with status code 403 the API keys will be renewed and the query will be tried a second time.

em92 commented:
> @em92 Do you got also 403 errors ?

Yes.

I will update new changes.

quickwick commented:
With the latest version of TwitterBridge.php from this branch(?), I am now getting "Bridge returned error 500! (19019)" and "Error message: 'Failed to make api call: $api returns 401'"

I've tried a variety of different twitter accounts, and did not select any of the checkmarks.

Edit: Previously (before the 4 most recent commits) I was getting a mix of "Bridge returned error 403!" and "Bridge returned error 401!"

arnd-s committed:
Fix error message

arnd-s commented:
The primary reasons for 401 or 403 errors are based on failed authentication. Can you try to drop you local cache ?

I'm adding some code for debug purposes now, to get more error details.  Twitter returns the full error reason inside the response body. The library function from lib/contents.php unfortunately throws an exception in case of a 4XX status code. So as an intermediate step, i will integrate the curl call inside the bridge and later extend lib/contents.php to have a version, which doesn't throw an exception.

arnd-s committed:
Implemented an own version the getContents functions from lib/contents.php to get extended error information.

This is an intermediate step to help debugging issues with the bridge without changes outside of the bridge. A patch for lib/contents.php to include a function like getContents but without throwing excetions will be added later

arnd-s committed:
Migrated the two missing contexts "by list" and "by list id"

mnalis commented:
Also, with version from https://github.com/RSS-Bridge/rss-bridge/pull/2433#issuecomment-1021171336, I've got one `500` error.

```
Bridge returned error 500! (19018)

Twitter @elonmusk was unable to receive or process the remote website's content!
Error message: `Requested username can't be found.`
Query string: `action=display&bridge=Twitter&context=By+username&u=elonmusk&nopic=on&format=Atom`
Version: `dev.2022-01-20`

    Press Return to check your input parameters
    Press F5 to retry
    Check if this issue was already reported on GitHub (give it a thumbs-up)
    Open a GitHub Issue if this error persists

arnd-s
```

other than that is seems to be working great so far! I'll update to latest state of PR now.

quickwick commented:
> The primary reasons for 401 or 403 errors are based on failed authentication. Can you try to drop you local cache ?

I've been deleting all the .cache files inside the cache folder, as well as the cache/TwitterBridge subfolder. Is there anything else that needs to be done.

With commit 7fd725f5, it seems to be working for most Twitter user feeds. However, there are still many that do not work. At first I thought the broken ones were limited to twitter Twitter accounts that contains "Potentially Sensitive Content", but that doesn't seem to always be the case:

Example:

> Twitter @NFLScenarios was unable to receive or process the remote website's content!
> Error message: `Failed to make api call: /1.1/statuses/user_timeline.json
> HTTP Status: 401
> Errormessage: {"request":"\/1.1\/statuses\/user_timeline.json","error":"Not authorized."}`
> Query string: `action=display&bridge=Twitter&context=By+username&u=NFLScenarios&format=Html`
> Version: `dev.2022-01-20`

arnd-s commented:
No, that should drop call caches. 

The error is indeed related to some problems with authentication. As far as i know, there is no difference between normal accounts and ones with  "Potentially Sensitive Content".
The latest version should in case of 401 or 403 error try to reget auth informations and then try again before given this error.

Maybe some rate limiting based on ip address ? How many accounts do you follow and how often do you poll them ?

Are you able to produce a debug trace for a poll with a 401 error and let me see it ? You just need to create a file named "DEBUG" inside root folder of rss-bridge. Then debug will be written to error_log of you webserver.
It would be best to have one clean example, without mixed in other polls.

quickwick commented:
> Maybe some rate limiting based on ip address ? How many accounts do you follow and how often do you poll them ?

I follow ~50 Twitter users. My feed reader is set to fetch articles once every 8 hours. From the rss-bridge web interface I am able to reproduce this issue by flipping back and forth between working and non-working Twitter users, so I don't think it's an IP rate limiting problem.

> Are you able to produce a debug trace for a poll with a 401 error and let me see it ? You just need to create a file named "DEBUG" inside root folder of rss-bridge. Then debug will be written to error_log of you webserver. It would be best to have one clean example, without mixed in other polls.

Here is the debug output from my lighttpd error log: https://pastebin.com/uPwh9Cx1
I hope I did that correctly.

linoth commented:
Been meaning to weigh in with an anecdote just so you've got the data point, and your question was the prompt that made me.

110 feeds, anywhere from hourly to daily with a few set to weekly, including a fair number of sensitive content users.  Probably mosting either 4 hours or 12 hours.  I don't think I've had a single auth issue since the PR started, and I've tried to pull your commits at least within 24 hours (pull origin pull/2433/head:).

Obviously something odd going on for some users, but once that's sorted out, the current updates seem reliable.

em92 commented on bridges/TwitterBridge.php:
```diff
@@ -613,19 +732,144 @@ private function getListId($username, $listName) {
-			return null;
+	private function getContents($url, $header = array(), $opts = array(), $returnHeader = false){
+		Debug::log('Reading contents from "' . $url . '"');
+
```
What is the difference with original getContents?

arnd-s replied:
Main reason is, that it don't throw a exception when http status code is >= 400.  Twitter adds a json coded error message inside the http response body. When using the version from the lib, this information is lost. 
This was a quick hack to prevent changes outside of twitterbridge. On long term, it would be better to adjust the lib version. 

I can adjust the lib version and add there a plain version, without status code evaluation and throwing exception. Extract the curl code into a separate function, which can be called directly and will be called from the original getContents function.
Or would it be better to make this in a separate PR ? I'm unfamiliar, what is preferred by other developers of the project.

em92 replied:
Made separate PR. https://github.com/RSS-Bridge/rss-bridge/pull/2447

em92 replied:
Suggesting following patch to adapt changes from 2447
https://pastebin.com/Z7ywpPDE

em92 replied:
Please ignore must patch. It is not valid. In any case, could you please adapt your bridge to latest changes in getContents that raises UnexpectedResponse exception?

em92 commented on bridges/TwitterBridge.php:
```diff
@@ -236,9 +240,66 @@ private function getApiURI() {
+			$params = array(
+				'slug'				=> strtolower($this->getInput('list')),
+				'owner_screen_name' => strtolower($this->getInput('user')),
+				'tweet_mode'		=> 'extended',
```
I suggest to use spaces between "=>" operator.

arnd-s replied:
Okay. I will convert them later

em92 commented on bridges/TwitterBridge.php:
```diff
@@ -236,9 +240,66 @@ private function getApiURI() {
+			$params = array(
+				'list_id'			=> $this->getInput('listid'),
+				'tweet_mode'		=> 'extended',
+			);
```
I suggest to use spaces between "=>" operator.

em92 commented on bridges/TwitterBridge.php:
```diff
@@ -251,94 +312,145 @@ public function collectData(){
+				if ($user->pinned_tweet_ids_str) {
+					$pinnedTweetId = $user->pinned_tweet_ids_str;
+				}
 			}
```
How about `if ($user && $user->pinned_tweet_ids_str)`?

arnd-s replied:
Sounds better ;)

em92 commented on bridges/TwitterBridge.php:
```diff
@@ -251,94 +312,145 @@ public function collectData(){
-		}, array());
+		// $promotedTweetIds = array_reduce($data->timeline->instructions[0]->addEntries->entries, function($carry, $entry) {
+		//	if (!isset($entry->content->item)) {
+		//		return $carry;
```
Instead of commenting lines, you should remove them. Better code quality.

arnd-s replied:
You are right. The reason was, that I wanted during the development to have the ability for taking a direct look on the old code from time to time when needed.

Before the final merge, i planed to make a whole cleanup to remove such parts. There are also functions and parameters which aren't used (anymore).

em92 commented on bridges/TwitterBridge.php:
```diff
@@ -251,94 +312,145 @@ public function collectData(){
-		if ($this->queriedContext === 'By username') {
-			foreach($data->timeline->instructions[0]->addEntries->entries as $tweet) {
-				if (!isset($tweet->content->item)) continue;
-				$tweetId = $tweet->content->item->content->tweet->id;
```
Why removing this?

arnd-s replied:
I haven't found the necessary elements in tweet structure from V1.1 API so far.
This is one case, i still need to migrate or at least proof, that it really isn't necessary anymore.

em92 replied:
I see. I suggest to remove it including commented lines.

arnd-s commented:
@quickwick 
> I hope I did that correctly.

I had only time taking a quick peek into the log. This was exactly what i wanted, thanks. I will later take a deep look and try to find out, if the code i wrote were working as expected or not and where the errors are coming from.

arnd-s committed:
Bugfix: Use string instead of integer/floating number for user id.

arnd-s commented:
@quickwick I think i found the bug, thanks to your logs. I have already added a commit, which should solve the problem. Would be nice, if you can give me feedback if it worked.

When there are no further flaws, i will concentrate on code cleanup and get it ready to be merged.

quickwick commented:
@arnd-s All my user feeds appear to be working once again with commit f1e3d21. Thank you!

There is a minor issue with images not showing up in retweets. From prior conversation in this thread, it's unclear to me whether this is a bug that needs fixing or expected behaviour with the 1.1 API.

Edit: After poking around a bit more, it seems the issue only occurs when the original tweet contained two or more images. Single-image retweets are working fine.

arnd-s committed:
Images and other media entities are now embedded on retweets

quickwick commented:
@arnd-s Commit c81ced4 seems to have taken care of the retweet media issue. Thank you!

Edit: Spoke too soon. There's a new issue with the retweets, perhaps specific to my feed reader. First, the retweets are no longer displaying as retweets. The Author value in my feed reader used to have "RT: " in front of the twitter handle, now it just shows the twitter handle. That's no big deal, what's more annoying is that the tweets are showing up again as unread every time I refresh a given feed in my reader.

After repeated tests, I believe the issue is limited to the situation where Twitter users are retweeting their own tweets. Retweets of other twitter users do not re-appear as unread items when refreshing a feed.

arnd-s commented:
I made some changes regarding retweets to fix missing images today.
I will push tomorrow morning some code cleanups, which should also include a fix for wrong author on retweets and solve the problem with read/unread tweets. I think, both points are related to each other.

arnd-s committed:
Code cleanup - reordered some codelines, removed old unused functions and code formating

arnd-s committed:
Fix phpcs error

quickwick commented:
The code changes in 769f8f3 / 83232d5 have not resolved the issue with self-retweets. They still do not show RT: in the Author field, and they re-appear as unread items on every feed refresh.

Edit: If you want an example, try the Twitter feed for @matt_obrien . He's a comedian. He has a tweet from yesterday (Feb 1) that he retweeted himself, something about eating spiders in our sleep. In my feed reader, that one tweet will keep reappearing each time I refresh the feed.

arnd-s committed:
Filter out retweets of users own tweets

arnd-s commented:
@quickwick Hope this solves the problem. The code now filters out the retweets from a user for his own original tweets.

BTW: Thanks for the concrete example. This helps development in general a lot.

quickwick commented:
@arnd-s I did some additional testing with 83232d5 and it looks like the self-retweet problem may be specific to my feed reader (RSS Guard). I was unable to reproduce with another feed reader (QuiteRSS). I will report this problem to the author of that program, we'll see what happens.

Thanks for the workaround in e95aac6 ! I have no desire to switch feed readers, and this was driving me crazy.

Edit: I may have figured out the problem. As part of submitting a bug report for RSS Guard, I had to capture a debug log. While reading through the debug log, it became obvious that the problem was happening because both entries in the feed (the original tweet, and the retweeted one) have the exact same id value. Is this something resolvable by you? It seems logical to me that id values should be unique.

Edit 2: I made this pastebin (https://pastebin.com/hTwKV6VP) originally for the RSS Guard bug report, but it's useful here. You can see there are two entries with the id value of "https:\/\/twitter.com\/matt_obrien\/status\/1488581884959363073"

Edit 3: According to the JSONfeed spec (https://www.jsonfeed.org/version/1/), "id (required, string) is unique for that item for that feed over time. If an item is ever updated, the id should be unchanged. New items should never use a previously-used id."

Akirainblack commented:
Sorry to ask what I'm sure is a stupid question, how could I go about testing this in docker?

Telkhine commented:
> Sorry to ask what I'm sure is a stupid question, how could I go about testing this in docker?

Here is the `docker-compose.yml` that I use:
```
version: '2'
services:
  rss-bridge:
    # image: rssbridge/rss-bridge:latest
    build: ./rss-bridge
    container_name: rss-bridge
    restart: unless-stopped
    volumes:
      - /usr/local/rss-bridge/config:/config
    ports:
      - 80:80
```

Then here are some of the commands you'll need to get it up and running:
```
# Stop the docker container
cd /usr/local/rss-bridge
docker-compose down

# Download the latest docker build from github
git clone --depth 1 https://github.com/RSS-Bridge/rss-bridge.git

# Download the latest TwitterBridge.php and copy it to the branch
wget https://raw.githubusercontent.com/RSS-Bridge/rss-bridge/e95aac6c79f19fa4feda8303fcc69f43c7972503/bridges/TwitterBridge.php
cp -v TwitterBridge.php rss-bridge/bridges/TwitterBridge.php


# Build the image
docker-compose build

# Start everything back up
docker-compose up -d
```

Modify this for your setup. You could also use git to merge in the branch, rather than primitive copying, but I'm a savage.

Akirainblack commented:
@Telkhine Thank you so much... I've now got twitter working. :D

quickwick commented:
Not sure what's gone wrong here, but TwitterBridge seems to be missing the majority of tweets. Have verified with twitter handles IGN, Steam, NintendoAmerica. When generating a feed, it's showing between 1/6 and 1/10 of the tweets, seemingly at random. Doesn't matter whether any filters are checked or not.

I tried clearing all the cache files, made no difference. Is anyone else seeing this?

Edit: Still happening the next morning. Grabbed a DEBUG output from lighttpd error log, in case it helps: https://pastebin.com/DXZXnqkB

arnd-s commented:
I will take a look at the problem and try to find out, what is happening. Thanks for the log.

quickwick commented:
I've spent the last day or so playing with Twitter APIs. For my own personal project, Python based. I don't know PHP, or I'd be a lot more useful here.

Anyway, I've come to two tentative conclusions. The first is that the Twitter v1.1 API seems to be returning 404 for some tweets that definitely exist. So the issue I'm seeing of missing tweets may be a Twitter-side issue, and not a problem with TwitterBridge. In addition, I went back to previous commits of TwitterBridge and am seeing the exact same behaviour of missing tweets.

Second, it appears that the standard "public" Bearer token does not work with v2 API. Works great with 1.1 API, but I get unauthorized errors when calling v2 endpoints. I wish you the best of luck in your attempts to make TwitterBridge v2 compatible without needing to use a twitter developer account. Hopefully v1.1 doesn't go away anytime soon.

em92 commented:
> When generating a feed, it's showing between 1/6 and 1/10 of the tweets, seemingly at random. Doesn't matter whether any filters are checked or not.

@quickwick 
I suggest you to compare feeds returned by TwitterBridge from current master and TwitterBridge from this PR to have more exact cases. To supress error messages returned by TwitterBridge from master, you should add following to config.ini.php:

```
[error]
output = "http"
```

quickwick commented:
I was mistaken about the missing tweets on the IGN, Steam, NintendoAmerica, etc user feeds. There are no missing tweets, it's just that the tweets are all out of order (not chronological) in the feed. In the HTML preview, as well as in the JSON/ATOM. This is essentially a non-issue, as any decent feed reader should allow you to sort chronologically, it's just a bit strange when previewing a feed.

However, the other issue I identified using my own Python code and the v1.1 endpoint is definitely happening. On some other user feeds, many tweets are definitely missing. The same tweets that are missing in the rss-bridge feed give a 404 response to the API call in my code, and the tweets that appear in the rss-bridge feed give a 200 response in my code. Looks very much like a Twitter API issue.

StrangeRanger commented:
I don't mean to be a bother, but if this PR fixes the 429 and 403 problems with the Twitter Bridge, is there a possible ETA on when it'll be finished?

mnalis commented:
@StrangeRanger you can download raw `TwitterBridge.php` file yourself: https://github.com/arnd-s/rss-bridge/raw/migrate_to_api_1.1/bridges/TwitterBridge.php and replace it and check how it works for your instance. For me it seems to work quite well, although I've had few spurious `error 500` (but only two over period of a week).

mnalis commented:
And here is the error I still get occasionally:

```
[Bridge returned error 500! (19042)](https://github.com/RSS-Bridge/rss-bridge/)

Twitter @SpaceX was unable to receive or process the remote website's content!
Error message: `Requested username can't be found.`
Query string: `action=display&bridge=Twitter&context=By+username&u=SpaceX&noimgscaling=on&format=Atom`
Version: `dev.2022-01-20`
```

(that is with latest file in this PR as of today, i.e. https://github.com/arnd-s/rss-bridge/commit/e95aac6c79f19fa4feda8303fcc69f43c7972503)

But otherwise PR seems to work fine for me, it is just a minor annoyance now.  (I only do basic twitter username following for dozen or so users, though, so don't know if other TwitterBridge methods work ok)

Sieboldianus commented:
> > Sorry to ask what I'm sure is a stupid question, how could I go about testing this in docker?
> 
> Here is the `docker-compose.yml` that I use:
> ...
> Modify this for your setup. You could also use git to merge in the branch, rather than primitive copying, but I'm a savage.

This is a shorter route, if you just want to test it:

```
docker exec -it docker_rss-bridge_1 /bin/bash
cd /app/bridges
mv TwitterBridge.php _TwitterBridge.php
curl https://raw.githubusercontent.com/arnd-s/rss-bridge/migrate_to_api_1.1/bridges/TwitterBridge.php -o TwitterBridge.php
chown www-data:www-data TwitterBridge.php
# CTRL+D
```

linoth commented:
Had noticed that Twitter seemed to be experimenting with privacy controls, i.e. you must log in to see this tweet, but hadn't bothered to investigate until today.

Behavior on this branch is to still provide tweets that are public, but exclude the login required tweets.  Haven't had the gumption to dig out what TwitterBridge is requesting, to look at what it's receiving, how RSS-Bridge parses that down and interprets it, etc so I'm just going to assume that an anonymous search isn't even shown that those tweets exist, let alone the contents.  Probably not solvable without playing with API keys, etc which is getting beyond the scope of RSS-Bridge's intent.

quickwick commented:
> Behavior on this branch is to still provide tweets that are public, but exclude the login required tweets. Haven't had the gumption to dig out what TwitterBridge is requesting, to look at what it's receiving, how RSS-Bridge parses that down and interprets it, etc so I'm just going to assume that an anonymous search isn't even shown that those tweets exist, let alone the contents. Probably not solvable without playing with API keys, etc which is getting beyond the scope of RSS-Bridge's intent.

I noticed 3 weeks ago that the public dummy Bearer Token used by this bridge to access Twitter V1.1 API seems to have stopped returning some tweets. The API response straight up claims the tweets don't exist, even though I can see them in the browser. This is why I wrote a new Bridge that queries the Twitter V2 API, using a personal Bearer Token. Unfortunately this means having to sign up for a Twitter developer account, create a Project and App and get a Bearer Token, and put it into config.ini.php. Works fine for private self-hosted RSS-Bridge, not sure if/how that will be manageable for public RSS-Bridge instances.

Anyone who wants to try is welcome to check out #2471 .

em92 commented:
I have following account: https://twitter.com/___em92
It gives following error
```
Requested username can't be found.
Twitter @___em92
```
https://feed.eugenemolotov.ru/?action=display&bridge=Twitter&context=By+username&u=___em92&format=Html

mnalis commented:
@em92 FWIW,  my private instance (running version of twitter bridge in this PR) seems to work just fine with that username

mnalis commented:
searching by hashtag with latest `TwitterBridge.php` from this PR seems to fail with `error 500` if there were no recent posts with that hashtag.
For example, https://twitter.com/hashtag/osmhr?src=hashtag_click returns post normally, with last one in 2019.

Using rss-bridge with `?action=display&bridge=Twitter&context=By+keyword+or+hashtag&q=%23osmhr&format=Atom` however fails with:
```
[Bridge returned error 500! (19071)](https://github.com/RSS-Bridge/rss-bridge/)
 

Twitter search #osmhr was unable to receive or process the remote website's content!
Error message: `No results for this query.`
Query string: `action=display&bridge=Twitter&context=By+keyword+or+hashtag&q=%23osmhr&format=Atom`
Version: `dev.2022-01-20`

    Press Return to check your input parameters
    Press F5 to retry
    Check if this issue was already reported on [GitHub](https://github.com/RSS-Bridge/rss-bridge/issues?q=is%3Aissue+is%3Aopen+Twitter+Bridge) (give it a thumbs-up)
    Open a [GitHub Issue](https://github.com/RSS-Bridge/rss-bridge/issues/new?title=Twitter+search+%23osmhr+failed+with+error+500&body=Error+message%3A+%60No+results+for+this+query.%60%0AQuery+string%3A+%60action%3Ddisplay%26bridge%3DTwitter%26context%3DBy%2Bkeyword%2Bor%2Bhashtag%26q%3D%2523osmhr%26format%3DAtom%60%0AVersion%3A+%60dev.2022-01-20%60&labels=Bridge-Broken&assignee=arnd-s)[](https://github.com/RSS-Bridge/rss-bridge/issues/new?title=Twitter+search+%23osmhr+failed+with+error+500&body=Error+message%3A+%60No+results+for+this+query.%60%0AQuery+string%3A+%60action%3Ddisplay%26bridge%3DTwitter%26context%3DBy%2Bkeyword%2Bor%2Bhashtag%26q%3D%2523osmhr%26format%3DAtom%60%0AVersion%3A+%60dev.2022-01-20%60&labels=Bridge-Broken&assignee=arnd-s) if this error persists

arnd-s
```

If that is due to API limitation (and not a bug), would it be possible to simply get empty RSS feed instead of `error 500` @arnd-s?
As  those recurring `error 500` are polluting several of my feeds (which track rare but important updates from twitter). 

When I use more often updated hashtag such as `#OpenStreetMap`, rss-bridge seems to work fine.

Other then that, it seems to work for me (apart from occasionally getting `error 500` which [autorecovers](https://github.com/RSS-Bridge/rss-bridge/pull/2433#issuecomment-1046010502) by the next poll or two)

dvikan commented:
@arnd-s Is the work going good with this PR?

berenddeschouwer commented:
If it helps, I've been running this for a month, and it mostly works.

The only issue I've encountered are the "error 500"s.  Whether they pollute the feed depend on a setting in config.ini

StrangeRanger commented:
I've encountered several "error 500"s as well, but I haven't really encountered any other errors.

mnalis commented:
> The only issue I've encountered are the "error 500"s. Whether they pollute the feed depend on a setting in config.ini


Can you tell us what should be put in `/etc/rss-bridge/config.ini.php` in order so those do not pollute the feed, @berenddeschouwer ?
That would be very helpful!

berenddeschouwer commented:
> > The only issue I've encountered are the "error 500"s. Whether they pollute the feed depend on a setting in config.ini
> 
> Can you tell us what should be put in `/etc/rss-bridge/config.ini.php` in order so those do not pollute the feed, @berenddeschouwer ? That would be very helpful!

under [error], change output to output=http

It depends on your rss client having a http client that deals with the '500' and tries again later.  This is usually true.

yamanq commented:
> > The only issue I've encountered are the "error 500"s. Whether they pollute the feed depend on a setting in config.ini
> 
> Can you tell us what should be put in `/etc/rss-bridge/config.ini.php` in order so those do not pollute the feed, @berenddeschouwer ? That would be very helpful!

You can also edit [report_limit](https://github.com/RSS-Bridge/rss-bridge/blob/master/config.default.ini.php#L81=) under `[error]` which you can raise to a higher number. Then errors will need to repeat multiple times before they are pushed to the feed.

em92 commented on bridges/TwitterBridge.php:
```diff
@@ -251,104 +272,136 @@ public function collectData(){
 
-			// Skip pinned tweet
+			// // Skip promoted tweets
+			// if (in_array($tweet->id_str, $promotedTweetIds)) {
```
Instead of commenting lines out, just remove them.

em92 commented on bridges/TwitterBridge.php:
```diff
@@ -251,104 +272,136 @@ public function collectData(){
+		// $promotedTweetIds = array_reduce($data->timeline->instructions[0]->addEntries->entries, function($carry, $entry) {
+		//    if (!isset($entry->content->item)) {
+		//        return $carry;
+		//    }
```
You can remove this including commented lines. Promoted tweets does not seem to appear.

em92 commented:
> I have following account: https://twitter.com/___em92 It gives following error
> 
> ```
> Requested username can't be found.
> Twitter @___em92
> ```
> 
> https://feed.eugenemolotov.ru/?action=display&bridge=Twitter&context=By+username&u=___em92&format=Html

False alert here.

em92 commented:
I am almost ready to merge this. I have left comments that need to be resolved before merge: https://github.com/RSS-Bridge/rss-bridge/pull/2433/files
Not sure if @arnd-s is active yet. If he does not answer next 7 days, feel free anyone to continue his work in separate PR.

Bockiii commented:
@em92 if the thing holding us up are your reviews on the "instead of comment out, remove", we can just do that in github ui, merge the recommendations and be done with it.

Bockiii commented on bridges/TwitterBridge.php:
```diff
@@ -251,104 +272,136 @@ public function collectData(){
+		//	}
+		// }
+
+		// Create output array with all required elements for each tweet
```
```suggestion

```

Bockiii commented on bridges/TwitterBridge.php:
```diff
@@ -251,104 +272,136 @@ public function collectData(){
 		// Extract tweets from timeline property when in username mode
 		// This fixes number of issues:
 		// * If there's a retweet of a quote tweet, the quoted tweet will not appear in results (since it wasn't retweeted directly)
 		// * Pinned tweets do not get stuck at the bottom
```
```suggestion

```

Bockiii commented on bridges/TwitterBridge.php:
```diff
@@ -251,104 +272,136 @@ public function collectData(){
+			// Skip spurious retweets
+			// if (isset($tweet->retweeted_status) && substr($tweet->text, 0, 4) === 'RT @') {
+			//	continue;
+			// }
```
```suggestion

```

Bockiii commented on bridges/TwitterBridge.php:
```diff
@@ -251,104 +272,136 @@ public function collectData(){
+			//	continue;
+			// }
+
+			// // Skip pinned tweet
```
```suggestion
		// Skip pinned tweet
```

Bockiii reviewed (commented):
Remove all the commented out lines.

Bockiii committed:
comment removal 1

Bockiii committed:
comment removal 2

Bockiii committed:
comment removal 3

Bockiii committed:
comment removal 4

Bockiii committed:
Update examplevalues

Bockiii commented:
There we go. I removed all commented lines that were not needed and updated the examplevalues. The "by hashtag" had a wrong syntax in it's examplevalue and the "by list" had no list name at all. "by list" now retrieves the same list id as the "by list id" one.

@em92 you can merge this if it pleases you now :)

Bockiii committed:
Merge branch 'master' into migrate_to_api_1.1

Bockiii commented on bridges/TwitterBridge.php:
```diff
@@ -252,104 +273,99 @@ public function collectData(){
+		//        $carry[] = $tweet->id;
+		//    }
+		//    return $carry;
+		// }, array());
```
```suggestion

```

Bockiii commented on bridges/TwitterBridge.php:
```diff
@@ -252,104 +273,99 @@ public function collectData(){
 
-			// Skip pinned tweet
+		// Skip pinned tweet
 			if ($hidePinned && $tweet->id_str === $pinnedTweetId) {
```
```suggestion
			// Skip pinned tweet
			if ($hidePinned && $tweet->id_str === $pinnedTweetId) {
```

Bockiii reviewed (approved):
linting and fixing

Bockiii committed:
more removals

Bockiii committed:
lint fixing

Bockiii commented on bridges/TwitterBridge.php:
```diff
@@ -252,104 +273,89 @@ public function collectData(){
+		}
+
 		$hidePictures = $this->getInput('nopic');
 
```
```suggestion
```

Bockiii commented on bridges/TwitterBridge.php:
```diff
@@ -252,104 +273,89 @@ public function collectData(){
+			if (isset($tweet->retweeted_status) && $tweet->retweeted_status->user->id_str === $tweet->user->id_str) {
 				continue;
 			}
 
```
```suggestion
```

Bockiii reviewed (commented):
Empty line fixing

Bockiii committed:
empty line 1

Bockiii committed:
empty line 2

Bockiii commented on bridges/TwitterBridge.php:
```diff
@@ -93,7 +93,7 @@ class TwitterBridge extends BridgeAbstract {
 				'name' => 'List',
 				'required' => true,
-				'exampleValue' => 'Tech-News',
+				'exampleValue' => 'Tech News',
```
```suggestion
				'exampleValue' => 'Tech-News',
```

Bockiii committed:
mistake on my end. typo

Bockiii commented:
Okay... NOW its fine :)

arnd-s commented:
Thanks @Bockiii 

Sorry for not responding the last weeks. I was rather busy and occupied. I will adept the getContents call to the new exception based handling from lib/contents.php until tomorrow.

If their are no further problems or blocking points, a merge would be good. New PRs can then be made for further fixes/tweaks.

arnd-s committed:
Change to use new version of getContents with improved exceptions from global lib instead own internal privat method

arnd-s commented:
I removed the private method getContents and make use of the new exception class.

Bockiii commented:
Alright, I think we are fine. Thanks to everyone involved in writing and testing!
